Add bank verificator

spxiwh opened the pull request:
This adds the bank verificator code to origin. Some things to highlight/comment on:

 * I have allowed banksim to run with a smaller delta_f than the waveform length. This allows banksim to run much faster.
 * Note how the filter-func-file option works in pycbc_banksim_match_combine. I've always said that I hate providing complex equations on the command line, so here's my alternative: Provide a python script that itself defines a function with specified input arguments and name. My "filter-func-file" that I've been using for hyperbank actually queries sbank directly to ask if a point would be rejected or not, but that script is not part of the pull request (maybe something for pycbc-config?)
 * I added the new-style colorbars to future.py. We do not require a matplotlib version that has these, but we should use these, so I add them (most of that was stolen from Alex Nielsen).
 * Version argument added to numerous codes.
 * No more defining post-processing scripts in the setup workflow(!!!!)
 * Banksim is *currently* still writing to XML, but all post-processing in done in HDF ... Hopefully XML can be abandoned here as well at some point! (We at least need to be able to use HDF banks!)
 * Stop overriding a builtin in render.py and ensure that directories exist in that function.

spxiwh commented:
A note on landscape's concerns: We are using the built-in pegasus properties often enough now that we probably shouldn't be declaring these as protected (underscore) variables. Should we change that?

ahnitz commented:
@spxiwh Ian, the right thing to do is a to make a public method that we can call instead (perhaps the same one).

titodalcanton commented on bin/plotting/pycbc_banksim_plot_fitting_factors:
```diff
@@ -0,0 +1,174 @@
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+""" Plot effective fitting factor from a bunch of point-source files.
```
The "effective" here looks incorrect for this script.

titodalcanton commented on bin/plotting/pycbc_banksim_table_point_injs:
```diff
@@ -0,0 +1,107 @@
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+""" Plot effective fitting factor from a bunch of point-source files.
```
I guess this also needs updating.

titodalcanton commented on bin/plotting/pycbc_banksim_table_point_injs:
```diff
@@ -0,0 +1,107 @@
+ 
+opt = parser.parse_args()
+
+col_names = ['Mass 1', 'Mass 2', 'Signal <br\\> recovery <br\\> fraction',
```
We probably want more quantities in this table: total mass, chirp mass, q, spins… Maybe configurable from the command line?

spxiwh replied:
I agree that this is useful, but I don't want to add more functionality within this pull request. Can we get this merged and then think about adding more information to result pages?

titodalcanton commented on bin/pycbc_banksim_combine_banks:
```diff
@@ -0,0 +1,64 @@
+__program__ = "pycbc_banksim_combine_banks"
+
+# Read command line options
+usage = """usage: %prog [options]"""
```
Is `usage` actually used?

titodalcanton commented on bin/pycbc_banksim_combine_banks:
```diff
@@ -0,0 +1,64 @@
+                    help="verbose output")
+parser.add_argument("-I", "--input-files", nargs='+',
+                    help="Explicit list of input files.")
+parser.add_argument("-o", "--output-file",  help="Output file name")
```
Please make this arg required.

titodalcanton commented on bin/pycbc_banksim_match_combine:
```diff
@@ -0,0 +1,181 @@
+lsctables.use_in(mycontenthandler)
+
+# Read command line options
+usage = """usage: %prog [options]"""
```
Is usage used?

titodalcanton commented on bin/pycbc_banksim_match_combine:
```diff
@@ -0,0 +1,181 @@
+#                    help="Explicit list of injection files. These must be in "
+#                         "the same order, and match one-to-one with the "
+#                         "match-files.")
+parser.add_argument("-o", "--output-file",  help="Output file name")
```
Please mark as required arg.

spxiwh commented:
@titodalcanton Please see response to comments made above, and associated patch.

titodalcanton commented:
Hi @spxiwh, unless @ahnitz has more comments, I'm ok with merging this and improving it later on.

ahnitz commented:
I've looked over the library changes, and I'm OK with them.

titodalcanton commented on bin/pycbc_banksim:
```diff
@@ -374,7 +374,7 @@ with ctx:
                                       filter_N, options.filter_sample_rate)
 
-            h_norm = sigmasq(htilde, psd=psd, low_frequency_cutoff=f_lower,
+                h_norm = sigmasq(htilde, psd=psd, low_frequency_cutoff=f_lower,
```
@spxiwh on a last pass, I noticed this change. I think the "missing" indentation is actually needed to enable the `--filter-high-freq-cutoff-isco-factor` option. Could you double check before I merge?

spxiwh replied:
I don't see a problem with this change. I don't see any input to this function that is set within the loop over signals. The flower is set in the outer for loop over templates and set to None before looping over signals. What do you think is the issue here?

The problem with the old method is that sigmasq is computed for *every* signal+template combination, but its input only depends on template parameters. This is not an insignificant contribution to computational cost, so should not be calculated when not needed.

titodalcanton replied:
Yeah, `f_lower` is constant within the loop, but `signal_high_freq_cutoff` is signal-dependent (when using `--filter-high-freq-cutoff-isco-factor`). I should have commented on the next line, sorry!

titodalcanton replied:
Note that I don't actually have a strong opinion on the usefulness of `--filter-high-freq-cutoff-isco-factor`. But if we *do* decide to break it in favor of the optimization, then let's just remove the whole option and related code.

spxiwh replied:
I'm not sure I understand this option at all? Is there a strong motivation to having this on master? Can I just remove it?

titodalcanton replied:
If I remember well it was added/used by @alex-nielsen for some studies related to NSBH signals with tidal disruption, where the early cutoff of the signal was approximated with a simple truncation. Let me check with him just in case.

spxiwh replied:
I've removed this option in the current commit.

spxiwh commented:
Hi @titodalcanton I added some commits to remove the high-frequency ISCO thing (I'm not sure that this code should be a "do everything" banksim ... We can make additional banksim executables if we want to do special things that are slower than the standard thing ... That would probably require moving a lot of this to a module to avoid code duplication and is outside the scope of this pull request).

I also added a new option in workflow to allow a list of input options that are *not* files (weird that this functionality didn't already exist). Then I used that to add links to the output tables in section 1, so you can go straight to followup on any interesting point injection:

https://www.atlas.aei.uni-hannover.de/~spxiwh/LSC/aLIGO/O2/template_banks/bank_verificator/O2_hyperbank_new/hdf_35geom_tito_bank/1._point_injection_sets/
